### PR TITLE
8292051: jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java failed "AssertionError: Expected terminated values: [666] but got: []"

### PR DESCRIPTION
**Pull request by AlanBateman**
Change the test to use a custom scheduler so there is better control on when the carrier threads terminates. Also convert test to TestNG so that all test cases can be run in the context of a virtual thread.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292051](https://bugs.openjdk.org/browse/JDK-8292051): jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java failed "AssertionError: Expected terminated values: [666] but got: []"

### Reviewers
 * [Peter Levart](https://openjdk.org/census#plevart) (@plevart - **Reviewer**) ⚠️ Review applies to [126584bc](https://git.openjdk.org/jdk/pull/9819/files/126584bc269e31f3168ffebb7a41930144658249)

### Contributors
 * Peter Levart `<plevart@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9819/head:pull/9819` \
`$ git checkout pull/9819`

Update a local copy of the PR: \
`$ git checkout pull/9819` \
`$ git pull https://git.openjdk.org/jdk pull/9819/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9819`

View PR using the GUI difftool: \
`$ git pr show -t 9819`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9819.diff">https://git.openjdk.org/jdk/pull/9819.diff</a>

</details>

### CODE REVIEW

File: test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java [hunk 1]
```diff
@@ -111,77 +118,57 @@ protected T initialValue() {
...
+
+            if (!terminatedValues.isEmpty()) {
+                throw new AssertionError("Unexpected terminated values after virtual thread terminated: " +
+                                         terminatedValues);
+            }
         }
 
+        // wait for carrier to terminate
+        Thread carrier = carrierRef.get();
+        carrier.join();
```
Comment (plevart) [inline test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java:145]: ...would `close()`-ing the `pool` already guarantee that all pool thread(s) are already stopped? `join()` is not needed then. But it doesn't hurt.

**Review by plevart — CHANGES_REQUESTED** [review]
Now that we have control over termination of carrier thread which can be stopped at will without waiting for 30s or more, we can again test all cases even with virtual threads...

**Comment by AlanBateman** [comment]
> Now that we have control over termination of carrier thread which can be stopped at will without waiting for 30s or more, we can again test all cases even with virtual threads...

Yes, we can. I'll convert the test to run as a TestNG test and we can use a data provider for that.

File: test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java (hunk 1 above)
Comment (AlanBateman) [reply test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java:145]: The close waits for the all tasks to finish. There is a small window between task completion and the thread termination.

File: test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java [hunk 2]
```diff
@@ -30,47 +30,67 @@
...
+            testCase(NULL, ttl -> { ttl.set(112); ttl.remove();               }      ),
+            testCase(NULL, ttl -> { ttl.set(112); ttl.remove(); ttl.get();    }, NULL),
+            testCase(NULL, ttl -> { ttl.get();    ttl.remove(); ttl.set(112); }, 112 ),
+
+            testCase(42, ttl -> { ttl.get();     ttl.set(NULL);               }, NULL),
+            testCase(42, ttl -> { ttl.set(NULL);                              }, NULL),
+            testCase(42, ttl -> { ttl.set(NULL); ttl.remove();                }      ),
+            testCase(42, ttl -> { ttl.set(NULL); ttl.remove(); ttl.get();     }, 42  ),
+            testCase(42, ttl -> { ttl.get();     ttl.remove(); ttl.set(NULL); }, NULL),
+        };
```
Comment (plevart) [inline test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java:84]: Hm, what about expanding the test cases programmatically. Like this?
https://github.com/plevart/jdk/commit/ffc681f3cccfdf473d4cc037e5f3e50f8172eec3
...it's easier to follow. Some nits included...

Comment (AlanBateman) [reply test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java:84]: > Hm, what about expanding the test cases programmatically. Like this?
> [plevart@ffc681f](https://github.com/plevart/jdk/commit/ffc681f3cccfdf473d4cc037e5f3e50f8172eec3)
> ...it's easier to follow. Some nits included...

I'm not sure that it's any easier to follow but I don't mind merging that in. The other cleanup is good.

**Pull request closed without merging**